packages/controllers/src/fsm.ts

```diff
@@ -0,0 +1,96 @@
+  InterpreterStatus,
+  StateMachine,
+  Typestate,
+} from '@xstate/fsm';
+import { assert } from './utils';
+
+/**
+ * Validates the set-up of a @xstate/fsm machine.
+ *
+ * 1. Ensures that all named actions in the config have a provided implementation.
+ *
+ * @param machine - The machine to validate.
+ * @throws {@link AssertionError}. If the validation fails.
+ */
+export function validateMachine<
+  TContext extends object,
+  TEvent extends EventObject,
+  TState extends Typestate<TContext>,
+>(machine: StateMachine.Machine<TContext, TEvent, TState>) {
+  assert('_options' in machine, 'The machine is not an @xstate/fsm machine');
+  const typed = machine as StateMachine.Machine<TContext, TEvent, TState> & {
+    _options: { actions?: StateMachine.ActionMap<TContext, TEvent> };
+  };
+
+  // 1.
+  const toArray = <T>(obj: T | T[]): T[] => {
+    if (Array.isArray(obj)) {
+      return obj;
+    } else if (obj === undefined || obj === null) {
+      return [];
+    }
+    return [obj];
+  };
+  const allActions = new Set<string>();
+  const addActions = (actions: any) =>
+    toArray(actions)
+      .flatMap<string>((action) => {
+        if (typeof action === 'string') {
+          return [action];
+        }
+        assert(typeof action === 'function');
+        return [];
+      })
+      .forEach(allActions.add.bind(allActions));
+
+  for (const state of Object.values<typeof typed.config.states[string]>(
+    typed.config.states,
+  )) {
+    addActions(state.entry);
+    addActions(state.exit);
+    for (const transition of Object.values<any>(state.on ?? {})) {
+      addActions(transition.actions);
+    }
+  }
+
+  allActions.forEach((action) =>
+    assert(
+      typed._options.actions !== undefined && action in typed._options.actions,
+      `Action "${action}" doesn't have an implementation`,
+    ),
+  );
+}
+
+/**
+ * Ensure that the interpreter is strict.
+ * Strict means that the transition must occur.
+ * The event must exist in .on {} state config and it's guard must succeed.
+ *
+ * The error will be thrown when an invalid `interpreter.send()` is called
+ * and will be bubbled there.
+ *
+ * TODO(ritave): Doesn't support self transitions.
+ *
+ * @param interpreter - The interpreter that will be force into strict mode.
+ * @throws {@link Error} Thrown when the transition is invalid.
+ */
+export function forceStrict(interpreter: StateMachine.Service<any, any, any>) {
+  // As soon as a listener subscribes, it is called. It might be called in
+  // an initial state which doesn't have the .changed property
+  let onInitialCalled = false;
+  interpreter.subscribe((state) => {
+    assert(!onInitialCalled || state.changed, 'Invalid state transition');
+    onInitialCalled = true;
+  });
+
+  const ogSend = interpreter.send.bind(interpreter);
+  interpreter.send = (...args) => {
+    assert(
+      interpreter.status === InterpreterStatus.Running,
+      'Interpreter is stopped',
+    );
+    return ogSend(...args);
+  };
+}
```

Comment:
This seems like a defensive programming measure. Which specific problems are we trying to avoid with this?

Reply:
xstate doesn't fail if the transition doesn't go through, it will just stay in the original state silently.

1. If no transition from current state exists for given event
2. If all guards fail for all transitions.

In our case those situations mean that there is some kind of wrong assumption or logic error happening inside `SnapController` and so failing fast is our best option to avoid security problems.